subprocess_utils.hpp

```diff
@@ -0,0 +1,219 @@
+#include <sys/wait.h>
+#include <sys/prctl.h>
+#include <cstring>
+
+
+namespace subprocess {
+    
+    /**
+     * A TwoWayPipe that allows reading and writing between two processes
+     * must call initialize before being passed between processes or used
+     * */
+class TwoWayPipe {
+	private:
+		//[0] is the output end of each pipe and [1] is the input end of each pipe
+		int input_pipe_file_descriptor[2];
+		int output_pipe_file_descriptor[2];
+        std::string internalBuffer;
+        bool inStreamGood = true;
+        
+	public:
+		TwoWayPipe() = default;
+		
+		/**
+		 * initializes the TwoWayPipe the pipe can not be used until this is called
+		 * */
+		void initialize(){
+			pipe(input_pipe_file_descriptor);
+			pipe(output_pipe_file_descriptor);
+		}
+	private:
+	    /**
+	     * closes the ends that aren't used (do we need to do this?
+	     * */
+		void closeUnusedEnds() {
+			//we don't need the input end of the input pipe or the output end of the output pipe
+			close(input_pipe_file_descriptor[1]);
+			close(output_pipe_file_descriptor[0]);
+		}
+
+	public:
+	
+	    /**
+         * sets this to be the child end of the TwoWayPipe
+         * linking the input and output ends to stdin and stdout/stderr
+         * */
+		bool setAsChildEnd() {
+		    int tmp[2] = {input_pipe_file_descriptor[0],input_pipe_file_descriptor[1]};
+
+		    input_pipe_file_descriptor[0] = output_pipe_file_descriptor[0];
+			input_pipe_file_descriptor[1] = output_pipe_file_descriptor[1];
+			output_pipe_file_descriptor[0] = tmp[0];
+			output_pipe_file_descriptor[1] = tmp[1];
+		    
+			dup2(input_pipe_file_descriptor[0], STDIN_FILENO);
+			dup2(output_pipe_file_descriptor[1], STDOUT_FILENO);
+			dup2(output_pipe_file_descriptor[1], STDERR_FILENO);
+
+			closeUnusedEnds();
+		}
+        
+        /**
+         * sets this pipe to be the parent end of the TwoWayPipe
+         * */
+		bool setAsParentEnd() {
+			closeUnusedEnds();
+		}
+        
+        /**
+         * writes a string to the pipe 
+         * @param input - the string to write
+         * @return the number of bytes written
+         * */
+		size_t writeP(const std::string &input) {
+			return write(output_pipe_file_descriptor[1], input.c_str(), input.size());
+		}
+
+        /**
+         * @return true unless the last call to read either failed or reached EOF
+         * */
+        bool isGood(){
+            return inStreamGood;
+        }
+
+        /**
+    	* reads up to n bytes into the internal buffer
+    	* @param n - the max number of bytes to read in
+    	* @return the number of bytes read in, -1 in the case of an error
+    	* */
+        ssize_t readToInternalBuffer()
+        {
+            char buf[256];
+            int cnt;
+            ssize_t bytesCounted = -1;
+
+            while((bytesCounted = read(input_pipe_file_descriptor[0], buf, 256))<=0){
+            	if (bytesCounted < 0) {
+            	    if (errno != EINTR){ /* interrupted by sig handler return */
```

Comment:
?

Reply:
read returns -1 on error. The specific error number is stored in the magic global variable "errno" all errors are unrecoverable except EINTR. This occurs when a signal interrupts the call and is not a reflection of a bad pipe. so try again, there are other codes that we may want to try again with but that's the only one that occurred.

Reply:
unix is icky